File: src/ru/gurps/generator/desktop/models/Character.java
package ru.gurps.generator.desktop.models;

import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import ru.gurps.generator.desktop.config.Model;
import ru.gurps.generator.desktop.models.characters.*;
import ru.gurps.generator.desktop.models.rules.*;

public class Character extends Model {
    public Integer id;
    public String player;
    public String name;
    public String currentPoints;
    public String maxPoints;
    public Integer st = 10;
    public Integer dx = 10;
    public Integer iq = 10;
    public Integer ht = 10;
    public Integer hp = 10;
    public Integer will = 10;
    public Integer per = 10;
    public Integer fp = 10;
    public Double bs = 5.0;
    public Integer move = 5;
    public Integer sm = 0;
    public Integer growth = 0;
    public Integer weight = 0;
    public Integer age = 0;
    public Integer tl = 0;
    public Integer tlCost = 0;
    public Integer head = 0;
    public Integer torse = 0;
    public Integer arm = 0;
    public Integer leg = 0;
    public Integer hand = 0;
    public Integer foot = 0;
    public Boolean noFineManipulators = false;
    
    public Character(){
    }

    public Character(String name, String maxPoints) {
        this.name = name;
        this.maxPoints = maxPoints;
    }

    public String getName() {
        return name;
    }

    public String getCurrentPoints() {
        return currentPoints;
    }

    public String getMaxPoints() {
        return maxPoints;
    }

    public ObservableList<Feature> features(){
        ObservableList<CharactersFeature> charactersFeatures = this.hasMany(new CharactersFeature());
        ObservableList<Feature> features = FXCollections.observableArrayList();
        for(CharactersFeature charactersFeature : charactersFeatures){
            Feature feature = (Feature) new Feature().find(charactersFeature.featureId);
            feature.add = true;
            feature.oldLevel = charactersFeature.level;
            feature.cost = charactersFeature.cost;
            features.add(feature);
        }
        return features;
    }

    public ObservableList<Skill> skills(){
        ObservableList<CharactersSkill> charactersSkills = this.hasMany(new CharactersSkill());
        ObservableList<Skill> skills = FXCollections.observableArrayList();
        for(CharactersSkill charactersSkill : charactersSkills){
            Skill skill = (Skill) new Skill().find(charactersSkill.skillId);
            skill.add = true;
            skill.level = charactersSkill.level;
            skill.cost = charactersSkill.cost;
            skills.add(skill);
        }

        ObservableList<CharactersSpecialization> charactersSpecializations = this.hasMany(new CharactersSpecialization());
        for(CharactersSpecialization charactersSpecialization : charactersSpecializations){
            Specialization specialization = (Specialization)
                    new Specialization().find(charactersSpecialization.specializationId);
            Skill skill = (Skill) new Skill().find(specialization.skillId);
            skill.name = skill.name + " (" + specialization.name + ")";
            skill.nameEn = skill.nameEn + " (" + specialization.nameEn + ")";
            skill.level = charactersSpecialization.level;
            skill.cost = charactersSpecialization.cost;
            skills.add(skill);
        }
        return skills;
    }

    public ObservableList<Spell> spells(){
        ObservableList<CharactersSpell> charactersSpells = this.hasMany(new CharactersSpell());
        ObservableList<Spell> spells = FXCollections.observableArrayList();
        for(CharactersSpell charactersSpell : charactersSpells){
            Spell spell = (Spell) new Spell().find(charactersSpell.spellId);
            spell.add = true;
            spell.level = charactersSpell.level;
            spell.finalCost = charactersSpell.cost;
            spells.add(spell);
        }
        return spells;
    }

    public ObservableList<Language> languages(){
        ObservableList<CharactersLanguage> charactersLanguages = this.hasMany(new CharactersLanguage());
        ObservableList<Language> languages = FXCollections.observableArrayList();
        for(CharactersLanguage charactersLanguage : charactersLanguages){
            Language language = (Language) new Language().find(charactersLanguage.languageId);
            language.spoken = charactersLanguage.spoken;
            language.written = charactersLanguage.written;
            language.cost = charactersLanguage.cost;
            languages.add(language);
        }
        return languages;
    }

    public ObservableList<Cultura> cultures(){
        ObservableList<CharactersCultura> charactersCulturas = this.hasMany(new CharactersCultura());
        ObservableList<Cultura> culturas = FXCollections.observableArrayList();
        for(CharactersCultura charactersCultura : charactersCulturas){
            Cultura cultura = (Cultura) new Cultura().find(charactersCultura.culturaId);
            cultura.cost = charactersCultura.cost;
            culturas.add(cultura);
        }
        return culturas;
    }

    public ObservableList<Quirk> quirks(){
        ObservableList<CharactersQuirk> charactersQuirks = this.hasMany(new CharactersQuirk());
        ObservableList<Quirk> quirks = FXCollections.observableArrayList();
        for(CharactersQuirk charactersQuirk : charactersQuirks){
            Quirk quirk = (Quirk) new Quirk().find(charactersQuirk.quirkId);
            quirk.cost = charactersQuirk.cost;
            quirks.add(quirk);
        }
        return quirks;
    }

    public String getPlayer() {
        if(player == null || player.equals("")) return "-";
        else return player;
    }

    public boolean delete(){
        new CharactersFeature().delete_all(new CharactersFeature().where("characterId", id));
        new CharactersModifier().delete_all(new CharactersModifier().where("characterId", id));
        new CharactersLanguage().delete_all(new CharactersLanguage().where("characterId", id));
        new CharactersSkill().delete_all(new CharactersSkill().where("characterId", id));
        new CharactersSpecialization().delete_all(new CharactersSpecialization().where("characterId", id));
        new CharactersCultura().delete_all(new CharactersCultura().where("characterId", id));
        new CharactersQuirk().delete_all(new CharactersQuirk().where("characterId", id));
        new CharactersSpell().delete_all(new CharactersSpell().where("characterId", id));
        new CharactersTechnique().delete_all(new CharactersTechnique().where("characterId", id));
        return super.delete();
    }
}

File: src/main/kotlin/ru/gurps/generator/desktop/models/Character.kt
package ru.gurps.generator.desktop.models

import javafx.collections.FXCollections
import javafx.collections.ObservableList
import ru.gurps.generator.desktop.config.Model
import ru.gurps.generator.desktop.models.characters.*
import ru.gurps.generator.desktop.models.characters.transports.CharactersTransportsAir
import ru.gurps.generator.desktop.models.characters.transports.CharactersTransportsGround
import ru.gurps.generator.desktop.models.characters.transports.CharactersTransportsSpace
import ru.gurps.generator.desktop.models.characters.transports.CharactersTransportsWater
import ru.gurps.generator.desktop.models.rules.*
import ru.gurps.generator.desktop.singletons.Characters
import ru.gurps.generator.desktop.singletons.Money
import java.sql.ResultSet
import java.util.*

class Character : Model {
    override var id: Int = -1
    var player: String = ""
    var name: String = ""
    var currentPoints: Int = 0
    var maxPoints: Int = 0
    var st: Int = 10
    var dx: Int = 10
    var iq: Int = 10
    var ht: Int = 10
    var hp: Int = 10
    var will: Int = 10
    var per: Int = 10
    var fp: Int = 10
    var bs: Double = 5.0
    var move: Int = 5
    var sm: Int = 0
    var growth: Int = 0
    var weight: Int = 0
    var age: Int = 0
    var tl: Int = 0
    var tlCost: Int = 0
    var head: Int = 0
    var torse: Int = 0
    var arm: Int = 0
    var leg: Int = 0
    var hand: Int = 0
    var foot: Int = 0
    var noFineManipulators: Boolean = false
    var moneyTotal: Long = 0
    var moneyTotalDollars: Long
        get() = moneyTotal / 100
        set(value) {
            moneyTotal = value * 100
        }

    var inventoryCost: Long = 0

    var moneyLeft: String = ""
        get() = Money.toDollars(moneyTotal - inventoryCost)

    constructor()
    constructor(name: String, maxPoints: Int) {
        this.name = name
        this.maxPoints = maxPoints
    }
    constructor(result: ResultSet) {
        id = result.getInt("id")
        val playerRes = result.getString("player")
        if (playerRes != null) player = playerRes
        val nameRes = result.getString("name")
        if (nameRes != null) name = nameRes
        currentPoints = result.getInt("currentPoints")
        maxPoints = result.getInt("maxPoints")
        st = result.getInt("st")
        dx = result.getInt("dx")
        iq = result.getInt("iq")
        ht = result.getInt("ht")
        hp = result.getInt("hp")
        will = result.getInt("will")
        per = result.getInt("per")
        fp = result.getInt("fp")
        bs = result.getDouble("bs")
        move = result.getInt("move")
        sm = result.getInt("sm")
        growth = result.getInt("growth")
        weight = result.getInt("weight")
        age = result.getInt("age")
        tl = result.getInt("tl")
        tlCost = result.getInt("tlCost")
        head = result.getInt("head")
        torse = result.getInt("torse")
        arm = result.getInt("arm")
        leg = result.getInt("leg")
        hand = result.getInt("hand")
        foot = result.getInt("foot")
        noFineManipulators = result.getBoolean("noFineManipulators")
        moneyTotal = result.getLong("moneyTotal")
        inventoryCost = result.getLong("inventoryCost")
    }

    fun alchemies(): ObservableList<Alchemy> {
        val charactersAlchemy = hasMany(CharactersAlchemy()) as ObservableList<CharactersAlchemy>
        val alchemies = FXCollections.observableArrayList<Alchemy>()
        charactersAlchemy.forEach {
            val alchemy = CharactersAlchemy().find(it.itemId) as Alchemy
            alchemy.add = true
            alchemies.add(alchemy)
        }
        return alchemies
    }

    fun features(params: HashMap<String, Any> = hashMapOf()): ObservableList<Feature> {
        val charactersFeatures = hasMany(CharactersFeature(), params) as ObservableList<CharactersFeature>

        val features = FXCollections.observableArrayList<Feature>()
        charactersFeatures.forEach {
            val feature = Feature().find(it.itemId) as Feature
            feature.add = true
            feature.level = it.level
            feature.cost = it.cost
            features.add(feature)
        }
        return features
    }

    fun skills(): ObservableList<Skill> {
        val charactersSkills = hasMany(CharactersSkill()) as ObservableList<CharactersSkill>
        val skills = FXCollections.observableArrayList<Skill>()
        charactersSkills.forEach {
            val skill = Skill().find(it.itemId) as Skill
            skill.add = true
            skill.level = it.level
            skill.cost = it.cost
            skills.add(skill)
        }

        specializations().forEach {
            val skill = Skill().find(it.skillId) as Skill
            val rm = skills.find { currentSkill -> currentSkill.id == it.skillId }
            skills.remove(rm)

            skill.name = "${skill.name} (${it.name})"
            skill.nameEn = "${skill.nameEn} (${it.nameEn})"
            skill.level = it.level
            skill.cost = it.cost
            skill.add = true
            skills.add(skill)
        }

        return skills
    }

    fun specializations(params: HashMap<String, Any> = hashMapOf()): ObservableList<Specialization> {
        val charactersSpec = hasMany(CharactersSpecialization(), params) as ObservableList<CharactersSpecialization>

        val specializations = FXCollections.observableArrayList<Specialization>()
        charactersSpec.forEach {
            val specialization = Specialization().find(it.itemId) as Specialization
            specialization.add = true
            specialization.level = it.level
            specialization.cost = it.cost
            specializations.add(specialization)
        }

        return specializations
    }

    fun techniques(): ObservableList<Technique> {
        val charactersTechnique = hasMany(CharactersTechnique()) as ObservableList<CharactersTechnique>
        val techniques = FXCollections.observableArrayList<Technique>()
        charactersTechnique.forEach {
            val technique = Technique().find(it.itemId) as Technique
            technique.add = true
            technique.level = it.level
            technique.cost = it.cost
            techniques.add(technique)
        }

        return techniques
    }

    fun spells(): ObservableList<Spell> {
        val charactersSpells = hasMany(CharactersSpell()) as ObservableList<CharactersSpell>
        val spells = FXCollections.observableArrayList<Spell>()
        charactersSpells.forEach {
            val spell = Spell().find(it.itemId) as Spell
            spell.add = true
            spell.level = it.level
            spell.finalCost = it.cost
            spells.add(spell)
        }

        return spells
    }

    fun languages(): ObservableList<Language> {
        val charactersLanguages = hasMany(CharactersLanguage()) as ObservableList<CharactersLanguage>
        val languages = FXCollections.observableArrayList<Language>()
        charactersLanguages.forEach {
            val language = Language().find(it.itemId) as Language
            language.spoken = it.spoken
            language.written = it.written
            language.cost = it.cost
            languages.add(language)
        }

        return languages
    }

    fun cultures(): ObservableList<Cultura> {
        val charactersCulturas = hasMany(CharactersCultura()) as ObservableList<CharactersCultura>
        val culturas = FXCollections.observableArrayList<Cultura>()
        charactersCulturas.forEach {
            val cultura = Cultura().find(it.itemId) as Cultura
            cultura.cost = it.cost
            culturas.add(cultura)
        }

        return culturas
    }

    fun quirks(): ObservableList<Quirk> {
        val charactersQuirks = hasMany(CharactersQuirk()) as ObservableList<CharactersQuirk>
        val quirks = FXCollections.observableArrayList<Quirk>()
        charactersQuirks.forEach {
            val quirk = Quirk().find(it.itemId) as Quirk
            quirk.cost = it.cost
            quirks.add(quirk)
        }

        return quirks
    }

    fun shields(): ObservableList<Shield> {
        val characterShields = hasMany(CharactersShield()) as ObservableList<CharactersShield>
        val shields = FXCollections.observableArrayList<Shield>()
        characterShields.forEach {
            val shield = Shield().find(it.itemId) as Shield
            shield.count = it.count
            shield.add = true
            shields.add(shield)
        }

        return shields
    }

    override fun destroy() {
        CharactersAlchemy().destroy_all(CharactersAlchemy().where("characterId", id))
        CharactersFeature().destroy_all(CharactersFeature().where("characterId", id))
        CharactersLanguage().destroy_all(CharactersLanguage().where("characterId", id))
        CharactersSkill().destroy_all(CharactersSkill().where("characterId", id))
        CharactersSpecialization().destroy_all(CharactersSpecialization().where("characterId", id))
        CharactersCultura().destroy_all(CharactersCultura().where("characterId", id))
        CharactersQuirk().destroy_all(CharactersQuirk().where("characterId", id))
        CharactersSpell().destroy_all(CharactersSpell().where("characterId", id))
        CharactersTechnique().destroy_all(CharactersTechnique().where("characterId", id))
        CharactersEquipment().destroy_all(CharactersEquipment().where("characterId", id))
        CharactersManArmor().destroy_all(CharactersManArmor().where("characterId", id))
        CharactersNotManArmor().destroy_all(CharactersNotManArmor().where("characterId", id))
        CharactersMeleeWeapon().destroy_all(CharactersMeleeWeapon().where("characterId", id))
        CharactersMeleeWeaponRanged().destroy_all(CharactersMeleeWeaponRanged().where("characterId", id))
        CharactersShield().destroy_all(CharactersShield().where("characterId", id))
        CharactersGun().destroy_all(CharactersGun().where("characterId", id))
        CharactersGrenade().destroy_all(CharactersGrenade().where("characterId", id))
        CharactersTransportsAir().destroy_all(CharactersTransportsAir().where("characterId", id))
        CharactersTransportsGround().destroy_all(CharactersTransportsGround().where("characterId", id))
        CharactersTransportsWater().destroy_all(CharactersTransportsWater().where("characterId", id))
        CharactersTransportsSpace().destroy_all(CharactersTransportsSpace().where("characterId", id))
        CharactersAlchemy().destroy_all(CharactersAlchemy().where("characterId", id))
        super.destroy()
    }
}
